File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        k = k % nums.length;
        reverse(nums, 0, nums.length-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, nums.length-1);
    }
    
    private void reverse(int[] nums, int s, int m){
            
        for (int i=s, j=m; i< nums.length; i++,j--){
            if (i < j){
                int tmp = nums[j];
                nums[j] = nums[i];
                nums[i] = tmp;
            }
        }
        
    }
}

File: 0189-rotate-array/0189-rotate-array.go
func rotate(nums []int, k int)  {
    
    n := len(nums)

    rotated := make([]int, n)

    for i:=0; i < n; i++ {
        rotated[(i+k)%n] = nums[i];
    }

    for i:= 0; i <n; i++ {
        nums[i] = rotated[i]
    }

}